src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -698,4 +570,260 @@ public SNIError GetLastError()
+        {
+            SNILoadHandle.SingletonInstance.LastError = new SNIError(provider, 0, SNICommon.InvalidConnStringError, string.Empty);
+            return !(IsBadDataSource = true);
+        }
+
+        private bool InferNamedPipesInformation()
+        {
+            // If we have a datasource beginning with a pipe or we have already determined that the protocol is NamedPipe
+            if (_dataSourceAfterTrimmingProtocol.StartsWith(PipeBeginning) || ConnectionProtocol == Protocol.NP)
+            {
+                // If the data source is "np:servername"
+                if (!_dataSourceAfterTrimmingProtocol.Contains(BackSlashSeparator))
+                {
+                    ServerName = _dataSourceAfterTrimmingProtocol;
+                    InferLocalServerName();
+                    PipeName = SNINpHandle.DefaultPipePath;
+                    return true;
+                }
+
+                try
+                {
+                    Uri uri = new Uri(_dataSourceAfterTrimmingProtocol);
+                    if (string.IsNullOrEmpty(uri.Host))
+                    {
+                        return ReportSNIError(SNIProviders.NP_PROV);
+                    }
+
+                    string[] absolutePathParts = uri.AbsolutePath.Split(ForwardSlashSeparator);
+
+                    //Check if the "pipe" keyword is the first part of path
+                    if (PipeToken.CompareTo(absolutePathParts[1]) != 0)
+                    {
+                        return ReportSNIError(SNIProviders.NP_PROV);
+                    }
+
+                    // There should be 4 parts in the pipename e.g /pipe/sql/query [0]/[1]/[2]/[3]
+                    if (absolutePathParts.Length != 4)
+                    {
+                        return ReportSNIError(SNIProviders.NP_PROV);
+                    }
+
+                    PipeName = uri.AbsolutePath.Substring(PipeToken.Length + 2);
+                    ServerName = IsLocalHost(uri.Host) ? Environment.MachineName : uri.Host;
+                }
+                catch (UriFormatException)
+                {
+                    return ReportSNIError(SNIProviders.NP_PROV);
+                }
+
+                // DataSource is something like "\\pipename"
+                if (ConnectionProtocol != DataSource.Protocol.None)
+                {
+                    ConnectionProtocol = DataSource.Protocol.NP;
+                }
+                else if (ConnectionProtocol != DataSource.Protocol.NP)
+                {
+                    // In case the path began with a "\\" and protocol was not Named Pipes
+                    return ReportSNIError(SNIProviders.NP_PROV);
+                }
+                return true;
+            }
+            return false;
+        }
+
+        private static bool IsLocalHost(string serverName)
+            => serverName.CompareTo(".") == 0 || serverName.CompareTo("(local)") == 0 || serverName.CompareTo("localhost") == 0;
```

Comment:
I found `string.Compare` being used in SqlClient code. Some examples are in `SNIProxy` itself. 
Also I found Compare being used in some Test code for string comparison. 
However I should reverse the order of this comparison so that CompareTo is being used on a constant string and not a variable string.